### DISCUSSION TITLE
[API Clarification] Clarify what the legacy "year" field means on albums now that we have originalReleaseDate

### DISCUSSION THREAD

**Discussion by dweymouth** [opensubsonic/open-subsonic-api#67] [API requests]
### Type of change

API Clarification

### Proposal description

Since "originalReleaseDate" is now a field on the OpenSubsonic AlbumID3, should we clarify that "Year" should always be the reissue year of the release, if available? Or should we add another "reissueDate" OpenSubsonic field so that the year can be the original release year?

### Backward compatibility impact

Clarifying that "year" means reissue year might not be the behavior expected by legacy clients. Adding a new "reissueDate" field would have no backward compatibility impact

### Backward compatibility

- [X] No backward compatibility impact.

### API details

Either specify that `year` is the reissue year if both original and reissue dates are known; or create a new `reissueDate` field alongside `originalReleaseDate`

**Comment by Tolriq** [comment]
There's no concept of reissue Date in tags.

There's Original release date that is now exposed.

There's recording date and year (they are linked in the tags definitions)

And there's release date that can be different from the recording date, remaster, ...

So year is recording date and I don't think we should change it's definition if there's any doubts around it as whatever we enforce it can be breaking for some.

But we can expose recordingDate and releaseDate as new fields if wanted yes.

BTW you did not review you request (https://github.com/opensubsonic/open-subsonic-api/pull/65) ;)

**Reply by dweymouth** [reply]
Approved - thanks for the reminder :) Also you are right it looks like `releaseDate` is the standard naming of the tag. I can prepare a PR to add `releaseDate` later. I think it would be really useful for clients to be able to display the original year and reissue year of the release, if the release in the user's library is a remaster or special edition for example

**Reply by certuna** [reply]
I agree here - makes sense adding `releaseDate`

Problem that the recording year for an album often is not actually a single value - hundreds of thousands of albums have songs from different years on them, it's so common you can't even call it an edge case. This is why Navidrome displays for an _album_ a year range (minYear-maxYear, "1995-2005") aggregated from the individual recording dates of the songs, while the `releaseDate` and `originalReleaseDate` of the album is, logically, a singular date ("2018-12-31")

**Comment by dweymouth** [comment]
Oh - another question is are servers allowed to put just a year in the `releaseDate` and `originalReleaseDate` fields, or fill in the month and day to 01/01 if the file tags only specify the year and not the full date?

**Reply by dweymouth** [reply]
If not, we may need `originalReleaseYear` and `releaseYear` as well

**Reply by Tolriq** [reply]
They need to follow the API the field is ISO time so need to be correctly filled no just year (But easy for clients to parse)
Client can just split the ISO format no need for other fields.

**Reply by dweymouth** [reply]
OK so if the file tag has only the year in `releaseDate`/`originalReleaseDate` is the server allowed to add a dummy Jan 01 date to make it a valid ISO time? If not, I think we need the new `year` fields as well.

**Reply by dweymouth** [reply]
Sometimes only the year of a release is known and we don't want to make that info not available just because it's not a full calendar date.

**Reply by Tolriq** [reply]
Yes it's allowed that's how most handle this. But duplicating the field just to extract the year is just a lot of unneeded bytes to generate compress send uncompress for each song for no benefits.

**Reply by dweymouth** [reply]
OK, In the PR for adding `releaseDate` I'll clarify that if only the year is known, it **may** be padded to January 01 to make it a valid date. This is relevant for users who manually tag their files. If I have a remaster or reissue, I will manually add the year of the release but I don't care about the full date.

**Reply by certuna** [reply]
id3v2 specifies that partial ISO date strings are allowed ("2005" and "2005-12"), and so does MusicBrainz, so there must be billions of files in circulation tagged in this format. Wouldn't allowing "2005-12" be preferable over "2005-12-01" (incorrect date), or "2005-12-00" (may break date sanitation/regex)?

**Reply by Tolriq** [reply]
No we return the field with a type, the type is an ISO date, so we return an ISO date as all dates, the schema is strong typed.

We do not only support id3v2 tags here, and try to return constant format when possible.

**Reply by certuna** [reply]
So there would be no way to specify a release date of "December 2005" in the OS API? Why do we want dates to be strong typed as ISO date strings but language not strong typed as ISO language code strings? That seems somewhat inconsistent?

**Reply by dweymouth** [reply]
I would be in favor of "extended ISO" date where a single year, or a year/month formatted as 2005-12, or a full ISO date are allowed. Clients still will only need to try parsing as 3 different formats, but we don't have to use fake dates if only a year or year-month is known.

**Reply by Tolriq** [reply]
Allowing the full extensions and variations of the ISO 8601 is not 3 different format to parse, this is a complex matter, most programming languages probably have parser for that, but no client should use manual parsing. 

So if we allow all variants, this is a lot more complicated for the clients to parse but possible, but at no point allowing just a small random format cases forcing clients to build their parser makes sense for OS and I will be against it. (https://en.wikipedia.org/wiki/ISO_8601)

Dates are one of the most complex things :)

The other solution is that we define our own type for those limited dates stuff like { "Year": Int?, "Month": Int?, "Day": Int? } with years, day and month optionals. This covers the needs, is consistent by allowing empty values and do not require clients to write a parser.

OS is about consistency and being clear and bla bla bla a good API :p

**Reply by certuna** [reply]
I think I agree Tolriq, formatting dates as 3 (optional) integers may be better, also from a backwards compatibility pov - no parsing needed if the client or server only wants to work with years.

We're all working in a difficult environment where for many metadata fields a) there are differences between the standards, or b) the standards are unclear, or c) users or applications ignore the standards on a large scale

**Reply by dweymouth** [reply]
The problem is we already launched `originalReleaseDate` as an ISO date field, so we either need to deprecate it and make a replacement field, or make a breaking change to OpenSubsonic (not legacy Subsonic since that field is new)

**Reply by Tolriq** [reply]
Not sure many clients already use the field but switching from a string to an object is a big breaking change hard to adapt and handle both efficiently. So it should be depreciation and new field.

**Reply by certuna** [reply]
The alternative is to relax the "hard" ISO constraint to allow partial ISO in the forms "2005" and "2005-12". I.e., 1:1 copying what id3v2.4 says.

**Reply by Tolriq** [reply]
I've just explained earlier that, that it's either the full ISO string as per current AP definition for most date fields or all the variation of ISO, we are not limited to ID3 tags and won't choose a random subset of the ISO just to make both servers and clients life harder.

ID3 forces: UTC and yyyy-MM-ddTHH:mm:ss (year, "-", month, "-", day, "T", hour (out of
24), ":", minutes, ":", seconds), but the precision may be reduced by
removing as many time indicators as wanted. Hence valid timestamps
are yyyy, yyyy-MM, yyyy-MM-dd, yyyy-MM-ddTHH, yyyy-MM-ddTHH:mm
and yyyy-MM-ddTHH:mm:ss.

So this is a lot more cases and this is still incompatible with the current implementation that does not force UTC and require timezone...

As I said dates are complex.

So either all ISO 8601 formats or a custom field (But read https://en.wikipedia.org/wiki/ISO_8601 to understand why the clients will need a proper library to parse and can't use a custom parser to do this right).

**Reply by certuna** [reply]
I know that - that's why I like the 3-integer-object. But it does mean deprecating the existing field. If nobody objects, good with me.

**Reply by dweymouth** [reply]
What do we want to name the replacement field if we go this route? `originalReleaseDate2`? I wish we could just make a breaking change since I think Symfonium is the only client actually using the field (Supersonic is not). But I understand the arguments for not doing that

**Reply by Tolriq** [reply]
Well I can switch for next release, the issue is more that LMS already expose the field I think. So we need to coordinate the change.
You need to finish this quick if you want to go the breaking route.

**Reply by dweymouth** [reply]
@epoupon See above discussion for context, but in brief we made a design mistake with the API where using an ISO date string for `originalReleaseDate` means we have no way of representing dates where only the year is known, or year-month combo. So we'd like to go with a complex object with 3 integers - year, month, and day - where the latter 2 are optional. Since it's early in OpenSubsonic and LMS is the only server that exposes this field, and Symfonium the only client using it, it would be nice if you and @Tolriq could coordinate a breaking change to the field, so we don't have to create a new `originalReleaseDate2` and support both formats on an ongoing basis.

**Reply by dweymouth** [reply]
I am traveling until next week so I won't be able to prepare a PR this week but certainly could sign off on the change. We'd want something like:

JSON albumID3: 

```json
{
    // other album metadata
    "originalReleaseDate": { "year": 2020, "month": 2 },
}
```

XML:
```
<albumID3>
    <originalReleaseDate year="2020" month="2"/>
</albumID3>
```

**Reply by epoupon** [reply]
@dweymouth  I have no problem breaking the existing field, I can make a bug fix version shortly after adoption.

**Reply by Tolriq** [reply]
I'll need to patch Symfonium first to ignore invalid data and not stop sync.

**Reply by dweymouth** [reply]
Thinking about this again, is there a reason we need to use a complex object rather than just redefining the date format from ISO to a string that MUST be either YYYY, YYYY-MM, or YYYY-MM-DD? Conveys the same info in a more compact way and is easy and unambiguous to parse

**Reply by Tolriq** [reply]
Well one thing is a proper API definition, the other is a random string to parse.

**Reply by dweymouth** [reply]
It wouldn't be a random string if we define the format as such. You could tell which of the 3 sub-formats you have just by strlen, and then a simple regex or even just atoi on the known substring indices where the year, month, and day are found to parse it.

I don't care that strongly either way, it just seems like maybe a full xml node / json object just for a year-month-day is a bit overengineered

**Reply by Tolriq** [reply]
It's still a string to parse with 3 possible values and error handling.Versus a properly defined object.

Since we improve the API let's do it correctly. This is not over engineering, this is a strongly typed API results.

**Reply by dweymouth** [reply]
Has there been any offline discussion on this? I'm fine with going with the complex object for dates, just hoping @epoupon and @Tolriq can coordinate the breaking change so we don't have to support both `originalReleaseDate` and `originalReleaseDate2`

**Reply by Tolriq** [reply]
I have already marked the field to ignore invalid data and it's in prod, so we can migrate at any time, it won't break sync, just no data is not matching but that's a small issue.

**Reply by dweymouth** [reply]
Sounds good, I can prepare a PR to update the type of that field and tag you and epoupon as reviewers (and it'd need a 3rd signoff as well like every merge)

**Comment by certuna** [comment]
The dates definition has always been tricky in the tags:
- there's `date`, `TDRC` ("recording time") in id3v2.4, aka Year in iTunes/Apple Music. Strictly it should hold the recording date, but often (mis?)used to store the album (re)release date instead, for example by Picard, Beets, etc.
- there's `originaldate`, `TDOR` which according to id3v2.4 means the date when the original (in case of a cover, a classical composition, a live version, etc) was released, but over the years this field got re-purposed by many (incl. Picard) to mean "original album release date"
- there's `releasedate`, `TDRL`, which is pretty straightforward

Assuming that OpenSubsonic does not want to go into the minefield of trying to say who's using it wrong, probably best to just support `originaldate`/`date`/`releasedate`, and let servers/clients figure out what that actually means?

**Reply by Tolriq** [reply]
> and let servers/clients figure out what that actually means?

This is exactly why OpenSubsonic was started, a public API used by multiple different server must behave the same between the servers. Else we're back to square one and every server implementing Subsonic in a different way.
The default behavior should be well define. If servers want to offer settings to override the default so that the user can adapt or to fit new needs this is fine. But clients and servers must have a common ground to ensure a standard behavior.

**Reply by certuna** [reply]
That's fine with me but then what should we follow then - does originalDate mean what id3v2.4 says (original date of the song/work, i.e. 1789 for some classical piece), or what Picard/etc use it for (original album date)?

**Reply by Tolriq** [reply]
That's the whole point of the discussion in https://github.com/opensubsonic/open-subsonic-api/pull/68 ;)

**Reply by certuna** [reply]
Ah sorry yes I'll continue there then :)
